Add nlp example

Ankur3107 opened the pull request:
## Describe changes
I implemented/fixed _ to achieve _.

## Pre-requisites
Please ensure you have done the following:
- [X] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [ ] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/zenml-io/zenml?pullRequest=467) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/zenml-io/zenml?pullRequest=467) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/zenml-io/zenml?pullRequest=467) it.</sub>

AlexejPenner commented on examples/nlp/README.md:
```diff
@@ -0,0 +1,47 @@
+# Implementation of NLP algorithm using Zenml
+
```
We have recently refactored all our example README files. Would you mind adopting a similar style to our other READMEs? 
Additionally it would be great to have some context regarding the model and what it does, so that beginners in the NLP space can quickly grasp what the pipeline is all about.

AlexejPenner replied:
Additionally for our examples we usually have the code at the same level as the README,  could you also please remove the `token-classification`-subdirectory?

Ankur3107 replied:
Actually, I am planning to add more NLP examples i.e text classification, question answering etc, So I made a sub-dir. You suggest me what should I do?

AlexejPenner reviewed (changes requested):
Thanks so much for your contribution to zenml Ankur, we appreciate it greatly! An NLP pipeline was sorely missing from our examples. 

I have made a few Change Requests that mostly relate to formatting and documentation. Let us know if you need assistance on anything.

htahir1 commented:
@Ankur3107 Apart from @AlexejPenner requests I would ask you to please use the `format.sh` and `lint.sh` scripts in the `scripts` folder as explained in the [CONTRIBUTING.md](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md)

P.S. I also changed the base branch to develop in accordance with the contributing guidelines

Ankur3107 commented on examples/nlp/token-classification/pipeline.py:
```diff
@@ -0,0 +1,147 @@
+  label_all_tokens = True
+  num_train_epochs = 3
+
+@step
```
done

Ankur3107 commented on examples/nlp/token-classification/pipeline.py:
```diff
@@ -0,0 +1,147 @@
+  mlflow.tensorflow.autolog()
+  model.fit(
+    train_set,
+    #validation_data=validation_set,
```
Sorry, I was running this code on colab to test, I forget to remove it.

Ankur3107 commented on examples/nlp/token-classification/run_pipeline.py:
```diff
@@ -0,0 +1,14 @@
+train_eval_pipeline(importer=data_importer().with_return_materializers(DatasetMaterializer),
+                       tokenizer=tokenization().with_return_materializers(DatasetMaterializer),
+               trainer=trainer(TokenClassificationConfig(num_train_epochs=1)).with_return_materializers(HFModelMaterializer),
+               evaluator=evaluator()).run()
```
done

Ankur3107 commented:
@AlexejPenner Thanks for your feedback. Actually, I was doing coding in colab and just copy-paste, So mixed basic stuff.

I have updated accordingly, Please check.

Ankur3107 commented:
This makes sense to me, I was also thinking the same to put materializer into integration. I will request you for review when it is ready.

AlexejPenner commented:
> This makes sense to me, I was also thinking the same to put materializer into integration. I will request you for review when it is ready.

Looking forward to it :)

Ankur3107 commented:
@AlexejPenner
Now I am ready for review. Please check.

List of things, I have tried to incorporate:
- Added hugging face materializer
- Added two NLP examples sequence and token classification
- Updated Readme

AlexejPenner commented:
> `format.sh` and `lint.sh`

I'll get right into it. In the meantime you could run our formatter and linter again in order to pass the tests:

scripts/format.sh 
scripts/lint.sh

AlexejPenner commented on examples/huggingface/README.md:
```diff
@@ -0,0 +1,103 @@
+- Tokenize and Prepare dataset for training: Use pre-trained tokenizer to tokenize and encode dataset into ids along with labels
+- Build and Train Model: You can build model or use pre-trained models from huggingface transformers. Use encoded dataset to train model.
+- Evaluate: Evaluate model loss and accuracy
+
```
I would also think, that a tiny example of the task taken out of the dataset could make sense here

AlexejPenner commented on examples/huggingface/README.md:
```diff
@@ -0,0 +1,103 @@
+pip install zenml transformers datasets
+
+# install ZenML integrations
+zenml integration install mlflow
+zenml integration install tensorflow
```
```suggestion
zenml integration install mlflow tensorflow
```

AlexejPenner commented on examples/huggingface/run_pipeline.py:
```diff
@@ -0,0 +1,104 @@
+            epochs=epochs,
+            batch_size=batch_size,
+        )
+        pipeline = token_classification.token_classifier_train_eval_pipeline(
+            importer=token_classification.data_importer(
+                token_classification_config
+            ).with_return_materializers(HFDatasetMaterializer),
+            load_tokenizer=token_classification.load_tokenizer(
+                token_classification_config
+            ).with_return_materializers(HFTokenizerMaterializer),
+            tokenization=token_classification.tokenization(
+                token_classification_config
+            ).with_return_materializers(HFDatasetMaterializer),
+            trainer=token_classification.trainer(
+                token_classification_config
+            ).with_return_materializers(HFTFModelMaterializer),
+            evaluator=token_classification.evaluator(
+                token_classification_config
+            ),
```
```suggestion
        pipeline = token_classification.token_classifier_train_eval_pipeline(
            importer=token_classification.data_importer(
                token_classification_config
            ),
            load_tokenizer=token_classification.load_tokenizer(
                token_classification_config
            ),
            tokenization=token_classification.tokenization(
                token_classification_config
            ),
            trainer=token_classification.trainer(
                token_classification_config
            ),
            evaluator=token_classification.evaluator(
                token_classification_config
            ),
```

These aren't necessary, based on the output type the Materializers will automatically be used.

AlexejPenner commented on examples/huggingface/run_pipeline.py:
```diff
@@ -0,0 +1,104 @@
+            epochs=epochs,
+            batch_size=batch_size,
+        )
+        pipeline = sequence_classification.seq_classifier_train_eval_pipeline(
+            importer=sequence_classification.data_importer(
+                sequence_classification_config
+            ).with_return_materializers(HFDatasetMaterializer),
+            load_tokenizer=sequence_classification.load_tokenizer(
+                sequence_classification_config
+            ).with_return_materializers(HFTokenizerMaterializer),
+            tokenization=sequence_classification.tokenization(
+                sequence_classification_config
+            ).with_return_materializers(HFDatasetMaterializer),
+            trainer=sequence_classification.trainer(
+                sequence_classification_config
+            ).with_return_materializers(HFTFModelMaterializer),
+            evaluator=sequence_classification.evaluator(
+                sequence_classification_config
+            ),
```
```suggestion
        pipeline = sequence_classification.seq_classifier_train_eval_pipeline(
            importer=sequence_classification.data_importer(
                sequence_classification_config
            ),
            load_tokenizer=sequence_classification.load_tokenizer(
                sequence_classification_config
            ),
            tokenization=sequence_classification.tokenization(
                sequence_classification_config
            ),
            trainer=sequence_classification.trainer(
                sequence_classification_config
            ),
            evaluator=sequence_classification.evaluator(
                sequence_classification_config
            ),
```

Same as above

AlexejPenner commented on examples/huggingface/sequence_classification.py:
```diff
@@ -0,0 +1,171 @@
+    tokenizer: PreTrainedTokenizerBase,
+    datasets: DatasetDict,
+) -> DatasetDict:
+    """Tokenizer dataset into tokens and then convert into encoded ids"""
```
```suggestion
    """Tokenize dataset into tokens and then convert into encoded ids"""
```

AlexejPenner commented on examples/huggingface/README.md:
```diff
@@ -0,0 +1,103 @@
+# Implementation of NLP algorithms using Zenml
```
As the directory is called huggingface it would be nice to have a small introductory paragraph what huggingface is and how it helps.

AlexejPenner reviewed (changes requested):
Thanks so much for all your hard work, this is very cool and we're really excited to get your work into our next release.

I have just a few more tiny suggestions.

htahir1 commented on examples/huggingface/sequence_classification.py:
```diff
@@ -0,0 +1,171 @@
+        len(tokenized_datasets["train"]) // config.batch_size
+    ) * config.epochs
+    optimizer, _ = create_optimizer(
+        init_lr=2e-5,
```
Might be part of the config

htahir1 commented on examples/huggingface/sequence_classification.py:
```diff
@@ -0,0 +1,171 @@
+    # Convert tokenized datasets into tf dataset
+    train_set = tokenized_datasets["train"].to_tf_dataset(
+        columns=["attention_mask", "input_ids"],
+        shuffle=True,
```
Might be part of the config

htahir1 commented on examples/huggingface/sequence_classification.py:
```diff
@@ -0,0 +1,171 @@
+        shuffle=True,
+        batch_size=config.batch_size,
+        collate_fn=DataCollatorWithPadding(tokenizer, return_tensors="tf"),
+        label_cols="label",
```
Might be part of the config

htahir1 commented on examples/huggingface/sequence_classification.py:
```diff
@@ -0,0 +1,171 @@
+    # Convert into tf dataset format
+    validation_set = tokenized_datasets["test"].to_tf_dataset(
+        columns=["attention_mask", "input_ids"],
+        shuffle=False,
```
Might be part of the config

htahir1 commented on examples/huggingface/sequence_classification.py:
```diff
@@ -0,0 +1,171 @@
+        shuffle=False,
+        batch_size=config.batch_size,
+        collate_fn=DataCollatorWithPadding(tokenizer, return_tensors="tf"),
+        label_cols="label",
```
Might be part of the config

htahir1 commented on examples/huggingface/sequence_classification.py:
```diff
@@ -0,0 +1,171 @@
+
+    # Convert into tf dataset format
+    validation_set = tokenized_datasets["test"].to_tf_dataset(
+        columns=["attention_mask", "input_ids"],
```
Might be part of the config

htahir1 reviewed (changes requested):
Fantastic job @Ankur3107 ! I think this is a great addition to ZenML. I am jumping in with my own review, if you dont mind. Its mostly refactorings and small suggestions that hopefully are quick and easy.

However, I do have one bigger request: Currently, it seems like the following functions wont work on the cloud with a GCP, AWS, or Azure bucket:

* `Dataset.from_parquet` / `Dataset.to_parquet`
* `DatasetDict.load_from_disk` / `DatasetDict.save_to_disk`
* `model.save_pretrained` / `model.from_pretrained`

For the first two, I think we can leverage the built-in integrations of HuggingFace that they speak about here: https://huggingface.co/docs/datasets/v1.4.0/filesystems.html . As they are using the same libraries as us for blob storage, I think we can do osmething clever their in the future. But for now, maybe we can add some simple logic to detect whether `self.artifact.uri` is `s3://`, `gs://` or `az://` ? 

For the third (i.e. the model) I would simple use `zenml.io.fileio` to copy from a local temp path to `self.artifact.uri`. This would be simple enough for the model files.

Does that make sense?

Ankur3107 commented:
@htahir1 
I like the idea and it is making sense as well. I will work on this.

I would like to take some time because I am traveling for my office work, I have very limited free time for few days, But I will add whenever I get time. Please give me some time.

htahir1 commented:
@Ankur3107 Please take your time we can make it part of the release to happen in two weeks :-)

htahir1 commented:
@Ankur3107 How are we looking? We are making the release next week so thought id ask you if we would make progress on this PR by then? Thanks!

Ankur3107 commented:
@htahir1 Yes, I have started working on this. 

can you help me, how can I get this `fs` https://github.com/zenml-io/zenml/blob/905a60ea90a3279a019dd298114ce4b1f7a9fcc9/src/zenml/integrations/aws/io/s3_plugin.py#L37 object because I want to use here
`ds.save_to_disk(self.artifact.uri,fs=fs)`

AlexejPenner commented:
> @htahir1 Yes, I have started working on this.
> 
> can you help me, how can I get this `fs`
> 
> https://github.com/zenml-io/zenml/blob/905a60ea90a3279a019dd298114ce4b1f7a9fcc9/src/zenml/integrations/aws/io/s3_plugin.py#L37
> 
> object because I want to use here
> `ds.save_to_disk(self.artifact.uri,fs=fs)`

Hey Ankur, sorry for the delayed answer. We recently had some changes within our code that affected the s3_plugin so we wanted to make sure this was stable before coming back to you. Additionally we had some internal discussions regarding Artifact stores and how their FS could be directly accessed from within a Materializer.

As of right now we don't have the perfect solution for this but we're working on it.

In the meantime you could try a workaround. You could create a temporary path locally that is written to and then copy that file into the artifact store fs. This should work as a temporary solution until we have implemented the more longterm solution. Be aware that this could fal if the amount of data becomes huge.

For your materializer the pseudocode would look a bit like this:

```python
def handle_return(self, ds: Type[Any]) -> None:
    """Writes a Dataset to the specified filename.
    Args:
        Dataset: The Dataset to write.
    """
    super().handle_return(ds)
    ...
    ds.save_to_disk(tmp_path)
    fileio.copy(tmp_path, self.artifact.uri)
```

htahir1 commented:
@Ankur3107  Would appreciate an update on this. Anyway we could help out? Would love to see this as part of the next release

Ankur3107 commented:
@htahir1 I am good for review. Please review and let me know your feedback.

htahir1 commented:
@Ankur3107 There are some merge conflicts, could you kindly fix them and we run the tests again? Then ill do a review

AlexejPenner commented on examples/huggingface/README.md:
````diff
@@ -0,0 +1,145 @@
+
+- Load dataset: Load sequence-classification dataset in this case it is the `imdb` dataset
+```python
+    from datasets import load_dataset
+    datasets = load_dataset("imdb")
+    print(datasets['train'][0])
+
+    {'label': 0, # Sentiment label i.e. 0->Negative 1->Positive
+ 'text': 'I rented I AM CURIOUS-YELLOW from my video store because of all the controversy that surrounded it when it was first released in 1967. I also heard that at first it was seized by U.S. customs if it ever tried to enter this country, therefore being a fan of films considered "controversial" I really had to see this for myself.......'}
````
Could you remove the indent here and add some manual linebreak to the text as it renders in a long scroll?

AlexejPenner commented on src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py:
```diff
@@ -0,0 +1,53 @@
+        super().handle_return(ds)
+        temp_dir = TemporaryDirectory()
+        ds.save_to_disk(temp_dir.name)
+        fileio.copy_dir(
```
copy_dir has moved to zenml.io.utils

Ankur3107 replied:
done

AlexejPenner commented on src/zenml/integrations/huggingface/materializers/huggingface_pt_model_materializer.py:
```diff
@@ -0,0 +1,60 @@
+        super().handle_return(model)
+        temp_dir = TemporaryDirectory()
+        model.save_pretrained(temp_dir.name)
+        fileio.copy_dir(
```
copy_dir has moved to zenml.io.utils

Ankur3107 replied:
done

AlexejPenner commented on src/zenml/integrations/huggingface/materializers/huggingface_tf_model_materializer.py:
```diff
@@ -0,0 +1,60 @@
+        super().handle_return(model)
+        temp_dir = TemporaryDirectory()
+        model.save_pretrained(temp_dir.name)
+        fileio.copy_dir(
```
copy_dir has moved to zenml.io.utils

Ankur3107 replied:
done

AlexejPenner commented on src/zenml/integrations/huggingface/materializers/huggingface_tokenizer_materializer.py:
```diff
@@ -0,0 +1,54 @@
+        super().handle_return(tokenizer)
+        temp_dir = TemporaryDirectory()
+        tokenizer.save_pretrained(temp_dir.name)
+        fileio.copy_dir(
```
copy_dir has moved to zenml.io.utils

Ankur3107 replied:
done

AlexejPenner reviewed (changes requested):
Unfortunately our last release changed the location of the copy_dir() method. Once you have accounted for that, I think this would be ready to merge. 

Somehow our linter is also failing, I am not entirely sure what the issue is. I will try investigating it on the side as well.

Ankur3107 commented:
@AlexejPenner updated as per your feedback.

AlexejPenner commented:
> @AlexejPenner updated as per your feedback.

Thank you so much for your patience and diligence in this contribution. We greatly appreciate your work on this.

AlexejPenner reviewed (approved):
Let's merge this :)

Ankur3107 commented:
@AlexejPenner @htahir1 Thank you for the guidance and patience. Will love to contribute in future. I am loving zenml framework.

AlexejPenner commented:
Thats amazing to hear. We'd love to have your future contributions. Feel free to drop by on our slack to connect or chat https://zenml.io/slack-invite/

Ankur3107 commented:
@AlexejPenner Thanks, I have already connected on slack.
